Add DDF support for Tuya device time sync

HahnBenjamin opened the pull request:
Allows DDF specifications to parse and respond to  TUYA_MCU_SYNC_TIME for timesync on the device.
The function should only be contained once in the DDF per device.
The actual attribute is not evaluated.

could be used in ddf on subdevice like so:
 ```json
{
  "name": "config/localtime",
  "public": false,
  "parse": { "fn": "tuyatime" }
}
```

HahnBenjamin commented:
I can confirm that:
> "Tuya debug 1 : Time sync request"

does not show up in the logs.

I also stepped through the code with the debugger when a "TUYA_TIME_SYNCHRONISATION" comes in. 
The _handleTuyaClusterIndication_ function will return with the if-check on [line 119](https://github.com/HahnBenjamin/deconz-rest-plugin/blob/tuya_time/tuya.cpp#L119)

Am I correct to assume that, with the addition of ddf fn "tuyatime", the handleTuyaClusterIndication function should ignore all manged devices?

Smanar commented:
>I also stepped through the code with the debugger

Jut by curiosity, what is your stuff ^^ ?

You have better stuff than me so I trust you, but in my mind :
- all tuya request are managed by handleTuyaClusterIndication no by pass before this fonction.
- if the device have a DDF and it s not the tuya synchronisation, the code is bypassed to use DDF core
- else it use legacy code

>does not show up in the logs.

This is strange, the method used previously is waiting for the device itself send the tuya synchronisation request, and after that deconz answer.

>I also stepped through the code with the debugger when a "TUYA_TIME_SYNCHRONISATION" comes in.
The handleTuyaClusterIndication function will return with the if-check on line 119

Its not because the tuya request is made by a device that is not in the api ?

`  Sensor *sensorNode = getSensorNodeForAddressAndEndpoint(ind.srcAddress(), ind.srcEndpoint());`

It check according the MAC adress and the endpoint, why it have not selected your created sensor ?

HahnBenjamin commented:
Sorry for not getting back earlier, I was busy with private affairs. Additionally it has been a few years since I have been working with C++. I also noticed that I messed up the timesync code while transferring it between OS environments. But I corrected this already on my side. Still need to update the PR though.

> Jut by curiosity, what is your stuff ^^ ?
> You have better stuff than me

I did not use too much fancy tools in my perspective. I used VSCode in conjunction with the GDB Debugger and attached it to the running deCONZ process. This works pretty much out-of-the-box., when you are in Linux. But I took a few detours for my development environment. 

I mainly work with Windows on my side, so I started the initial work on a dedicated Ubuntu booting from an external drive. I got annoyed with the dual boot so I moved over to using WSL for compiling on Ubuntu on Windows. By the time I noticed the existence of gWSL on Windows 11. So I gave this a try. You can USB passthough the ConBee Stick to the WSL instance, so I was able to build and run the Ubuntu Variant on my Windows Laptop 👯. While checking the project file for the plugin I also noticed that all required files for building on Windows are included in the deCONZ Windows Installation Package. Given that I also spent some time to get a functional build on Windows running. 

> Its not because the tuya request is made by a device that is not in the api ? But it can explain why the debug line is never displayed.
> 
> ` Sensor *sensorNode = getSensorNodeForAddressAndEndpoint(ind.srcAddress(), ind.srcEndpoint());`
> 
> It check according the MAC adress and the endpoint, why it have not selected your created sensor ?
> 
> When you see the tuya synchronisation it come naturally from the device, or it s forced by deconz code ?

I was testing this on a dedicated deCONZ Zigbee network and hardware with the purpose of supporting the two tuya devices I ordered for evaluation. Only one device was joined the network while testing. The  tuya timesync command is sent by the device and also processed by the plugin. I am not sure if my .[ddf file for the sensor](https://github.com/HahnBenjamin/deconz-rest-plugin/blob/tuya_ts0601/devices/tuya/_TZE200_locansqn_temperature_humidity_sensor.json) device was malformed, but _getSensorNodeForAddressAndEndpoint_ check failed. I am not too familiar with the deCONZ architecture yet: I made a test by supplying a function to check against the MAC Adress instead of the _Adress_ class: `getSensorNodeForAddressAndEndpoint(ind.srcAddress().ext(), ind.srcEndpoint());`
That actually passed and and made the device sync time using the "legacy" code in the tuya.cpp.

It might have todo with a "fingerprint" entry for the ddf missing as I can see this being used by _getSensorNodeForAddressAndEndpoint_ for the equality check. I did not investigate this further. I also did not find a reference what the purpose or structure for the "fingerprint" field in the DDF files is. Is there any written information about it?
 
I am also wondering if there is some sort of roadmap for the "REST API plugin" as I understand the goal of the DDF transition is to cleanup and simplify the device mapping and maintainance. However if I see the code in the device_access_fn.cpp this already seems to mutate to some sort of godfile implementing all device mapping code in one place.

Smanar commented:
> While checking the project file for the plugin I also noticed that all required files for building on Windows are included in the deCONZ Windows Installation Package. Given that I also spent some time to get a functional build on Windows running.

There is windows files ?

`if (sensor.fingerPrint().endpoint != ep)                            { continue; }`

Ha yes, It can explain some issue I have too with endpoint (and BTW better to specify the endpoint, never use "auto")

Here you have a sample for a sensor using fingerprint https://github.com/dresden-elektronik/deconz-rest-plugin/blob/master/devices/bosch/bwa-1_water_sensor.json can too specify "out" cluster.
Don't need to put all clusters.

HahnBenjamin commented:
Closing the PR since the the tuya time sync for ddf devices was fixed in #6305
